SoyEtiago/Proyecto-EDyA2
# Add Firebase Authentication with Email and Password Support, Logging Middleware, and reorganize Server Setup

This PR introduces the following features and improvements to the application:

- Firebase-based user registration and login routes using email and password authentication.
- Express middleware for request logging that tracks requests, response times, and status codes.
- Server setup improvements including CORS handling and routes for authentication and home.

### Changes:

- Authentication (Auth) Routes:
        Added /register and /login routes for handling user registration and login using Firebase authentication.
        Input validation is handled using express-validator for secure and structured data handling.

- Request Logging Middleware (reqLog):
        Middleware captures important request details such as method, URL, client IP, and user agent.
        Logs the request start time and calculates the response time once the response is finished.
        Logs the status code for each request and outputs it in a structured format.

- Server Setup (app.js):
        Configured the server to listen on PORT from environment variables, with a fallback to 4000.
        Applied CORS to handle requests from specific allowed origins, restricting methods and headers for security.
        Integrated the logging middleware and the express.json() parser for handling incoming JSON data.
        Routed API endpoints to /api/auth and /api/home.

### Files Modified:

- authRoute.js:
        Handles user registration (POST /register) and login (POST /login) routes.
        Utilizes express-validator middleware for input validation.
- reqLog.js:
        Middleware for logging request details such as URL, method, response time, status code, and client information.
- app.js:
        Sets up the Express app with middleware, routing, CORS configuration, and starts the server.

Middleware:

- reqLog Middleware:
        Tracks requests and logs details about the client, request method, and URL.
        Logs response time and status after the response is finished, helping with performance tracking and debugging.

## How to Test:

1. Start the Server:
- Ensure you have the correct .env configuration for PORT and Firebase credentials.
- Run the server with npm run backend.
- Expected: Server should start and listen on the specified port.

2. Test Request Logging:
- Make requests to the API (/api/auth/login, /api/auth/register).
- Expected: Console should log the request details, including IP address, user agent, response time, and status.

3. Test Auth Routes:
- Test registering a new user with a valid email and password at /api/auth/register.
- Test logging in with the same credentials at /api/auth/login.
- Expected: Successful registration and login responses, with validation errors for incorrect input.

Screenshots:
![imagen](https://github.com/user-attachments/assets/ebdd88c5-74e4-4d13-9c7f-be5b82c31601)
![imagen](https://github.com/user-attachments/assets/e25e6d81-7162-430d-9599-f3e79322dceb)
![imagen](https://github.com/user-attachments/assets/92491f48-8adf-4ed3-a0d4-f71e10c21c5b)
![imagen](https://github.com/user-attachments/assets/12ad9243-b353-4b80-a620-5c9d50110676)

The reqLog middleware currently outputs logs to the console. If production-level logging is required, a logging service or library should be integrated (e.g., winston, morgan).
[Morgan](https://www.npmjs.com/package/morgan)
[Wingston](https://github.com/winstonjs/winston)

Next Steps:
- Integrate token-based authentication for session management.
